From 52484201feeed10e3d62cb7ef33194d879d2b374 Mon Sep 17 00:00:00 2001
From: Simon Moll <simon.m.moll@googlemail.com>
Date: Fri, 4 Apr 2014 11:19:37 -0300
Subject: [PATCH 1/2] instantiating FASan pass with address sanitizer

---
 lib/CodeGen/BackendUtil.cpp |    1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/CodeGen/BackendUtil.cpp b/lib/CodeGen/BackendUtil.cpp
index 90b0f68..e6bef90 100644
--- a/lib/CodeGen/BackendUtil.cpp
+++ b/lib/CodeGen/BackendUtil.cpp
@@ -173,6 +173,7 @@ static void addAddressSanitizerPasses(const PassManagerBuilder &Builder,
       static_cast<const PassManagerBuilderWrapper&>(Builder);
   const CodeGenOptions &CGOpts = BuilderWrapper.getCGOpts();
   const LangOptions &LangOpts = BuilderWrapper.getLangOpts();
+  PM.add(createSelectivePageMigrationPass());
   PM.add(createAddressSanitizerFunctionPass(
       LangOpts.Sanitize.InitOrder,
       LangOpts.Sanitize.UseAfterReturn,
-- 
1.7.9.5

